pkg/oc/clusteradd/components/docker-registry-operator/docker_registry_operator.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/openshift/origin/pkg/oc/clusterup/docker/dockerhelper"
+	"github.com/openshift/origin/pkg/oc/clusterup/manifests"
+)
+
+const (
+	namespace                = "openshift-core-operators"
+	RegistryServiceClusterIP = "172.30.1.1"
+)
+
+type DockerRegistryOperatorComponentOptions struct {
+	InstallContext componentinstall.Context
+}
+
+func (c *DockerRegistryOperatorComponentOptions) Name() string {
+	return "openshift-docker-registry-operator"
+}
+
+func (c *DockerRegistryOperatorComponentOptions) Install(dockerClient dockerhelper.Interface) error {
+	kubeAdminClient, err := kubernetes.NewForConfig(c.InstallContext.ClusterAdminClientConfig())
+	if err != nil {
+		return err
+	}
+
+	// create the actual resources required
+	imageTemplate := variable.NewDefaultImageTemplate()
+	imageTemplate.Format = c.InstallContext.ImageFormat()
+	imageTemplate.Latest = false
+
+	params := map[string]string{
+		"IMAGE":                 imageTemplate.ExpandOrDie("hypershift"),
+		"LOGLEVEL":              fmt.Sprintf("%d", c.InstallContext.ComponentLogLevel()),
+		"COMPONENT_IMAGE":       imageTemplate.ExpandOrDie("registry"),
+		"COMPONENT_LOGLEVEL":    fmt.Sprintf("%d", c.InstallContext.ComponentLogLevel()),
+		"NAMESPACE":             namespace,
+		"OPENSHIFT_PULL_POLICY": c.InstallContext.ImagePullPolicy(),
+	}
+
+	glog.V(2).Infof("instantiating registry-operator template with parameters %v", params)
+
+	component := componentinstall.Template{
+		Name:            "openshift-registry-operator",
+		Namespace:       namespace,
+		RBACTemplate:    manifests.MustAsset("install/openshift-registry-operator/install-rbac.yaml"),
+		InstallTemplate: manifests.MustAsset("install/openshift-registry-operator/install.yaml"),
+
+		// wait until there's an available endpoint of registry
+		WaitCondition: func() (bool, error) {
+			glog.V(2).Infof("polling for registry availability ...")
+			deployment, err := kubeAdminClient.AppsV1().Deployments("openshift-registry").Get("registry", metav1.GetOptions{})
+			if apierrors.IsNotFound(err) {
+				return false, nil
+			}
+			if err != nil {
+				return false, err
+			}
+			if deployment.Status.AvailableReplicas == 0 {
+				return false, nil
+			}
+			return true, nil
+		},
+	}
+	err = component.MakeReady(
+		c.InstallContext.ClientImage(),
+		c.InstallContext.BaseDir(),
+		params).Install(dockerClient)
+	if err != nil {
+		return err
+	}
+
+	// we to selectively add to the config, so we'll do this post installation.
+	operatorClient, err := operatorversionclient.NewForConfig(c.InstallContext.ClusterAdminClientConfig())
+	if err != nil {
+		return err
+	}
+	operatorConfig, err := operatorClient.DockerregistryV1alpha1().OpenShiftDockerRegistryConfigs().Get("instance", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	/*
+		masterPublicHostPort, err := getMasterPublicHostPort(c.InstallContext.BaseDir())
+		if err != nil {
+			return err
+		}
+	*/
+	// TODO: set REGISTRY_OPENSHIFT_SERVER_ADDR
```

Comment:
we actually don't want to set this on cluster up anyway because the client machine can't DNS resolve the service hostnames anyway, so we want the registry to use the service ip.